packages/ra-core/src/auth/useCanAccess.ts

```diff
@@ -0,0 +1,124 @@
+ * @param {Object} params.record Optional. The record to check access for
+ *
+ * @returns Return the react-query result and a canAccess property which is a boolean indicating the access status
+ *
+ * @example
+ *     import { useCanAccess } from 'react-admin';
+ *
+ *     const PostDetail = () => {
+ *         const { isPending, canAccess } = useCanAccess({
+ *             resource: 'posts',
+ *             action: 'read',
+ *         });
+ *         if (isPending || canAccess) {
+ *             return null;
+ *         }
+ *         return <PostEdit />;
+ *     };
+ */
+export const useCanAccess = <ErrorType = Error>(
+    params: UseCanAccessOptions<ErrorType>
+): UseCanAccessResult<ErrorType> => {
+    const authProvider = useAuthProvider();
+    const logoutIfAccessDenied = useLogoutIfAccessDenied();
+    const { onSuccess, onError, onSettled, ...queryOptions } = params ?? {};
+
+    const result = useQuery({
+        queryKey: ['auth', 'canAccess', JSON.stringify(params)],
+        queryFn: async ({ signal }) => {
+            if (!authProvider || !authProvider.canAccess) {
+                return true;
+            }
+            return authProvider.canAccess({
+                ...params,
+                signal,
+            });
+        },
+        ...queryOptions,
+    });
+
+    const onSuccessEvent = useEvent(onSuccess ?? noop);
+    const onSettledEvent = useEvent(onSettled ?? noop);
+    const onErrorEvent = useEvent(
+        onError ??
+            ((error: ErrorType) => {
+                if (process.env.NODE_ENV === 'development') {
+                    console.error(error);
+                }
+                logoutIfAccessDenied(error);
+            })
+    );
+
+    useEffect(() => {
+        if (result.data === undefined || result.isFetching) return;
+        onSuccessEvent(result.data);
+    }, [onSuccessEvent, result.data, result.isFetching]);
+
+    useEffect(() => {
+        if (result.isError === false) {
+            return;
+        }
+        onErrorEvent(result.error);
+    }, [onErrorEvent, result.error, result.isError]);
+
+    useEffect(() => {
+        if (result.status === 'pending' || result.isFetching) return;
+        onSettledEvent(result.data, result.error);
+    }, [
+        onSettledEvent,
+        result.data,
+        result.error,
+        result.status,
+        result.isFetching,
+    ]);
+
+    return useMemo(() => {
+        return {
+            ...result,
+            canAccess: result.data,
+        };
+    }, [result]);
+};
+
+export interface UseCanAccessOptions<ErrorType = Error>
+    extends Omit<UseQueryOptions<boolean, ErrorType>, 'queryKey' | 'queryFn'> {
+    resource: string;
+    action: string;
+    record?: unknown;
+    onSuccess?: (data: boolean) => void;
+    onError?: (err: ErrorType) => void;
+    onSettled?: (data?: boolean, error?: ErrorType | null) => void;
+}
+
+export type UseCanAccessResult<ErrorType = Error> = QueryObserverResult<
```

Comment:
Can we be a bit smarter and use a union to help TypeScript know when `canAccess` is actually non-null, just as we do in `useGetList`?